docs/architecture/adr-001-autopilot-oci-basic-auth-support.md

```diff
@@ -0,0 +1,137 @@
+## Decision
+
+Implement support in Autopilot for pulling artifacts, such as k0s binaries and image bundles, directly from a registry using the [ORAS](https://oras.land/docs/) client. Additionally, add support for HTTP authentication to ensure secure access to artifacts.
+
+## Solution
+
+Starting with the current `PlanResourceURL` struct:
+
+```go
+type PlanResourceURL struct {
+        // URL is the URL of a downloadable resource.
+        URL string `json:"url"`
+
+        // Sha256 provides an optional SHA256 hash of the URL's content for verification.
+        Sha256 string `json:"sha256,omitempty"`
+}
+```
+
+We must specify to Autopilot where to access credentials for remote artifact pulls. This will be achieved by adjusting the struct as follows:
+
+```go
+type PlanResourceURL struct {
+        // URL is the URL of a downloadable resource.
+        URL string `json:"url"`
+
+        // Sha256 provides an optional SHA256 hash of the URL's content for verification.
+        Sha256 string `json:"sha256,omitempty"`
+
+        // ArtifactPullSecrets holds a reference to a secret where the credentials are
+        // stored. We use these credentials when pulling the artifacts from the provided
+        // URL using any of the supported protocols (http, https, and oci).
+        ArtifactPullSecret *ArtifactPullSecret `json:"artifactPullSecret,omitempty"`
+
+        // Insecure indicates whether certificates in the remote URL (if using TLS) can
+        // be ignored, aka InsecureSkipVerify.
+        Insecure bool  `json:"insecure,omitempty"`
+}
+```
+
+The `ArtifactPullSecret` property will be added and its struct will be defined as follow:
+
+```go
+type ArtifactPullSecret struct {
+      // Namespace of the secret.
+      Namespace string `json:"namespace"`
+
+      // Name of the secret.
+      Name string `json:"name"`
+}
+```
+
+The secret pointed by the provided `ArtifactPullSecret` will be used for pulling artifacts using either HTTP[S] or OCI protocols and is expected to by of type `kubernetes.io/dockerconfigjson` if the protocol in use is `oci://` or of type `Opaque` if protocols `http://` or `https://` are used (see below for details on the Secret layout).
+
+Example configuration for OCI:
+
+```yaml
+url: oci://my.registry/binaries/k0s:v1.30.1+k0s.0
+sha256: e95603f167cce6e3cffef5594ef06785b3c1c00d3e27d8e4fc33824fe6c38a99
+artifactPullSecret:
+  namespace: kube-system
+  name: artifacts-registry
+```
+
+Example configuration for HTTPS:
+
+```yaml
+url: https://my.file.server/binaries/k0s-v1.30.1+k0s.0
+sha256: e95603f167cce6e3cffef5594ef06785b3c1c00d3e27d8e4fc33824fe6c38a99
+artifactPullSecret:
+  namespace: kube-system
+  name: artifacts-basic-auth
+```
+
+Example configuration for HTTP:
+
+```yaml
+url: http://my.file.server/binaries/k0s-v1.30.1+k0s.0
+sha256: e95603f167cce6e3cffef5594ef06785b3c1c00d3e27d8e4fc33824fe6c38a99
+artifactPullSecret:
+  namespace: kube-system
+  name: artifacts-token-based-auth
+```
+
+### Secrets Layout
```

Comment:
There's also the [TLS Secrets](https://kubernetes.io/docs/concepts/configuration/secret/#tls-secrets) secret type; mTLS could be a very interesting future addition.

Reply:
That is a nice idea. We can always create an addendum for this in the future. I need to avoid scope creep here. This proposal started as "oci + basic auth" and it is already "oci + basic auth + token".